### ISSUE TITLE
Unexpected behaviour of non-persisted has_many through associations

### ISSUE THREAD

**Issue by RKushnir** [rails/rails#9548]
Consider a basic hm:t association

```
class Host < ActiveRecord::Base
  has_many :links
  has_many :pages, through: :links
end
```

(`Link` and `Page` are trivial).

If you add and remove an item from `links`, it still thinks the item is magically present... in the empty collection:

```
host = Host.new
page = Page.new
host.pages.include? page
 => false
host.pages << page
host.pages.delete page
host.pages.include? page
 => true 
host.pages
 => #<ActiveRecord::Associations::CollectionProxy []>
```

Tried this on both 3.2.12 and 4.0.0.beta1.

PS. If you save the records, it works as expected.

**Comment by nerdinand** [comment]
I tried to reproduce this on `master`. It still seems to be a problem:
https://gist.github.com/nerdinand/7665496e2b6d26619781

**Comment by nerdinand** [comment]
Proposed patch still works. Should I send a new PR?

**Comment by kamipo** [comment]
Fixed by #29593.
